src/app_engine/analytics_page.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2015 Google Inc. All Rights Reserved.
+
+"""Module for the AnalyticsPage handler."""
+
+import json
+import time
+
+import webapp2
+
+import analytics
+import constants
+from constants import AnalyticsField
+
+
+class AnalyticsPage(webapp2.RequestHandler):
+  """Client Analytics data handler.
+
+  Each POST body to the AnalyticsPage is a JSON object of the form,
+  {
+    'request_time_ms': <client time in milliseconds when the request was sent>,
+    'type': <type of the content below>,
+    'content': <content related to the above type>
+  }
+
+  Currently the AnalyticsPage handles the following types:
+  'event': A JSON event object of the form:
+          {
+            'event_type': <string corresponding to an attribute of EventType>,
+            'event_time_ms': <client time when the event occurred>,
+            'room_id': <id of the room corresponding to the event [optional]>,
+          }
+
+      The event is reported to Analytics.report_event with the
+      timestamp of the event modified to take into consideration the
+      clock differences between the client and the
+      server. 'request_time_ms' is compared to server time to provide
+      an offset, and the recorded event 'timestamp' is 'event_time_ms'
+      adjusted by the calculated offset. The original 'event_time_ms'
+      is reported as the 'client_timestamp'. This ignores the latency
+      of opening a connection and sending the body of the message to
+      the Analytics page. Travel time should be negligible.
+  """
+
+  def _write_response(self, result, message=None):
+    response = {'result': result}
+    if message is not None:
+      response['message'] = message
+
+    self.response.write(json.dumps(response))
+
+  def _write_invalid_request(self):
+    self.error(400)
```

Comment:
If we update the other APIs we'll still want invalid request reason for them. So the response code would be 400 but we'd still have JSON response content with the specific error code. This seems to just add client / server parsing code unnecessarily.